### DISCUSSION TITLE
help for AUR package

### DISCUSSION THREAD

**Discussion by LuckyTurtleDev** [Robosturm/Commander_Wars#886] [Q&A] [answered]
I have started to create an AUR package for this games. see https://aur.archlinux.org/packages/commander-wars
current PKGBUILD:  https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=commander-wars

Sadly I still have some question:
* What does the `addr2line` executable do? Did I need it?
* What does the javascript files at the `templates` folder do? Did I need them?
* ~~Exist an icon, witch I can use for a desktop file?~~
* Exist cmake flags, witch I should use?
* Should the package version be `0.21.4` or `21.4`? (Beta_21_4 is not allowed)
* It is possible to change the save patch? The game save everything to the working directory. Instead it should use `~/.config/commander-wars` or `~/.local/share/commander-wars`

I have also notify that Command Wars does not follow the Linux/Unix Directory Structure. So current the files must be moved manually, after compiling.

**Comment by Robosturm** [comment, marked as answer]
* addr2line was used for stack-tracing on previous versions, but is currently unused and not needed. Even if stack-tracing should be readded it's a fully optional exe, which removes the stacktrace output in case of a crash
* templates folder is an optional folder containing scripts which can be used to automated certain things upon start or for manually creating scripts for a map.
* The ingame version is 0.21.4 so i would call it this way.
* in the ini-file created after the first start is an entry UserPath which can be changed to tell the game where it should store data. (Except for the ini-file itself). This is already used on android devices.
* You can find the current cmake setup used on all OS's here: https://github.com/Robosturm/Commander_Wars/blob/master/.github/workflows/release.yml
Which are:
        -B '${{github.workspace}}/build' \
        -DCMAKE_BUILD_TYPE=${{env.BUILD_TYPE}} \
        -DCMAKE_INSTALL_PREFIX='${{github.workspace}}/install' \
        -DCMAKE_PREFIX_PATH="${{github.workspace}}/qt/Qt/${{matrix.qtPath}}/lib/cmake/" \
        -DQt6_DIR=${{github.workspace}}/qt/Qt/${{matrix.qtPath}}/lib/cmake/Qt6/
And
    - name: Build
      working-directory: '${{github.workspace}}/build'
      run: cmake --build '${{github.workspace}}/build' --config $BUILD_TYPE --target install -j16
* All together there is no real required Flags needed
* Linux/Unix currently uses the Windows install directory structure since i have no clue what those conventions are. However the cmake can be adapted to support linux structure deployment. See https://github.com/Robosturm/Commander_Wars/blob/master/CMakeLists.txt line 830 and on going deal with the installation into the given install path.

**Reply by LuckyTurtleDev** [reply]
Thanks for the fast answer.

The files should be structure as follow:
* exeuctables at `bin`
* libarys at `lib/commander-wars/`
* platform independent data (as example audios, grafics, templates, maps) at `share/commander-wars/`
* desktop icon at `share/pixmaps/commander-wars.EXTENSION`
* `.desktop` (start menu entry) at `share/applications/Commander Wars.desktop`

All path without the `INSTALL_PATH` prefix.

I am not very familiar with cmake, but I can try to adjust some path and add a `.desktop` file. But I have sadly no idea, how to add new files to cmake. Is a pull request are welcome?

According to this line the current `platform` folder does only include the qt-libraries:
https://github.com/Robosturm/Commander_Wars/blob/57b2d286552585200c4d61972b23ae057703e290/CMakeLists.txt#L852
Why are they needed? Are they not include at the qt packages? I have delete them and I have not found a issue so far.


>in the ini-file created after the first start is an entry UserPath which can be changed to tell the game where it should store data. (Except for the ini-file itself). This is already used on android devices.

This is sadly not very useful, because this file can not create during for installation and must be created for all users. Furthermore if the user starts commander wars from a different working directory, the ini-file can not be found by the game.
Does not exist a c++ equivalent for the [directories crate](https://github.com/dirs-dev/directories-rs), which can give you the default data/config path at each operating system?
I think it is pretty annoying, if you start the games from the wrong folder and have a lot of auto generated files at your home folder, witch you must delete manual now. (Already happen to me multiple times)

**Reply by Robosturm** [reply]
* Cause on minimum windows on pc with not qt installed. Those libs are not always found/installed.
* Well that's a philosphy discussion. I prefer having the files in the path of the exe, so i don't need to search for a global file, somewhere in the orbit. 
I understand your point of view and i guess you can change that with a define option for the source code dealing with that.
In the settings.cpp is already an option for "small screen devices" e.g. android, who changes the default-path....

**Reply by LuckyTurtleDev** [reply]
> I prefer having the files in the path of the exe, so i don't need to search for a global file, somewhere in the orbit.

This is simple not possible for a package, because the user had no write access to `/usr/bin`.
But Commander Wars does not use the path of the executable, it does use the working directory.

I can workaround this issue, the same way I do it with proprietary gog games, by adding a start script, witch does change the working directory. But I do not prefer this option, because then I must install the executable to `/opt`. 
Simple changing the `Settings::m_settingFile` const at the `settings.cpp` should not work because it must evaluated at runtime. (I think `~` would not work as path). But qt provide a [function](https://doc.qt.io/archives/qt-4.8/qdesktopservices.html#storageLocation) for this. Maybe I will try to create a patch for this, witch I can apply at the PKGBUILD.


The file structure is also not a problem for the package, because it can be changed by the PKGBUILD after building.
But I think it would confused many user, if the try to build and install the game by themself.

**Reply by Robosturm** [reply]
I think you missunderstood me.
What i meant is the following:
Before the ini is actually created/or read in the settings.cpp
It creates a path at which the ini can be found  and where other things should be created.
This path could be changed during compile-time to use one of the other qt-standard paths described here:
https://doc-snapshots.qt.io/qt6-dev/qstandardpaths.html

However for windows having it in the same folder is the prevert way to go. Hence my suggestion using a define/cmake-flag

**Reply by LuckyTurtleDev** [reply]
Thanks to your help, I was able to create a [patch](https://aur.archlinux.org/cgit/aur.git/tree/data_location.patch?h=commander-wars) witch I can apply.
Sadly I did not know how `ifdef` can detect if a cmake flag is set. So I can not create a pull request for it.

I have update the [AUR package](https://aur.archlinux.org/packages/commander-wars). It does install Commader Wars to the system including start menu entry and icon.

**Reply by Robosturm** [reply]
You can checkout the branch https://github.com/Robosturm/Commander_Wars/tree/vectorWork
I added the cmake configure option USEAPPCONFIGPATH which sets a define which toggles the default path in the settings.cpp

**Reply by LuckyTurtleDev** [reply]
How can I select this opiton? I have add `-DUSEAPPCONFIGPATH=ON` to my cmake flags, but it seems not to work.

**Reply by LuckyTurtleDev** [reply]
`DUSEAPPCONFIGPATH` was hard coded.
It does now work after https://github.com/Robosturm/Commander_Wars/pull/893

Thanks for you help

**Comment by Robosturm** [comment]
@Lukas1818 changed the installation structure with a minor change using the Project-Name instead of commander-wars
You may take a look at this commit:
https://github.com/Robosturm/Commander_Wars/commit/e28f89cb822933e1b6db75773e816cd76a4ce664
I can't test if the strucuture matches your expectation but change it with the given code should be doable

**Reply by LuckyTurtleDev** [reply]
```
usr/
└── Commander_Wars_Release
    ├── bin
    │   └── Commander_Wars
    ├── lib
    │   └── Commander_Wars
    │       ├── addr2line
    │       └── platforms
    │           ├── libqeglfs.so
    │           ├── libqlinuxfb.so
    │           ├── libqminimalegl.so
    │           ├── libqminimal.so
    │           ├── libqoffscreen.so
    │           ├── libqvnc.so
    │           ├── libqwayland-egl.so
    │           ├── libqwayland-generic.so
    │           └── libqxcb.so
    └── share
        ├── Commander_Wars
        │   └── templates
        │       ├── gamescript.js
        │       └── init.js
        └── pixmaps
            └── Commander_Wars.png
                └── icon.png
```
Looks good, but there are some small issues.
I have create a pr for this: https://github.com/Robosturm/Commander_Wars/pull/899

**Comment by lonewolf19991** [comment]
This is the method I have used to install the game under Debian based distributions, which I wrote for reference: https://www.commanderwars.com/mediawiki/Compiling_on_Linux/
From what I have noticed, the bare minimum requirements are these:

`set(Qt6_DIR "/home/username/Qt/6.2.3/gcc_64/lib/cmake/Qt6") 
set(XDG_RUNTIME_DIR "/run/user/1000") 
set(XKB_INCLUDE_DIR "/usr/include/xcb") 
set(XKB_LIBRARY "/usr/lib/x86_64-linux-gnu/") 
set(Vulkan_INCLUDE_DIR "/usr/lib/x86_64-linux-gnu/")`

Granted, I have been meaning to grab the XDG runtime dir through a bash command instead of manually integrating it, but I myself have not dealt with CMake enough.

A package for a debian based system would be great but before that, all copyrighted material would need to be removed, as to avoid any lawyers.

**Reply by LuckyTurtleDev** [reply]
>A package for a debian based system would be great but before that, all copyrighted material would need to be removed, as to avoid any lawyers.

So far I can tell at current state that would mean, to replace nearly every grafic and sound. But I think this must be done at the future other wise the project will be shutdown some day. Especially after the release of the Advance Wars remake.

Luckily this is not an issue for the aur package, because it does only include build instruction and neither the source code or precompile binaries.

**Reply by LuckyTurtleDev** [reply]
> From what I have noticed, the bare minimum requirements are these:

I have never executed one of this `set()`s and it does compile without issues. see [PKGBUILD](https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=commander-wars)
Maybe they are needed if you does not use your package manager, for installing qt, gcc, etc. So I do not thing you need to execute them if you have installed the packages.

**Comment by lonewolf19991** [comment]
That's unusual. I never got it working without spoon-feeding those directories to CMake. In fact, I even had to spoon-feed it the latest version of gcc I had to install because of me having to compile it from source. 

That said, Debian version I am running only has qt 6.0.1 or so in the official repositories. 

Also,  until 6.2.3 there were audio issues with the compilation, seemingly because of a glitch with Qt, so it does amuse me that you don't need any of that. Tried it in both Fedora and Debian.

Anyway, regarding the situation of copyright you could ask around to check which units and content are fully created from scratch and they could probably be repackaged as software which can be set on a store or an official repository.
